maps/src/main/java/com/gluonhq/impl/maps/TileImageView.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2020, Gluon
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.impl.maps;
+
+import com.gluonhq.maps.tile.TileRetriever;
+import com.gluonhq.maps.tile.TileRetrieverProvider;
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.ReadOnlyBooleanWrapper;
+import javafx.beans.property.ReadOnlyDoubleProperty;
+import javafx.beans.property.ReadOnlyDoubleWrapper;
+import javafx.beans.property.ReadOnlyObjectProperty;
+import javafx.beans.property.ReadOnlyObjectWrapper;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Logger;
+
+public class TileImageView extends ImageView {
+
+    private static final Logger logger = Logger.getLogger(TileImageView.class.getName());
+    private static final Image placeHolder = new Image(TileImageView.class.getResourceAsStream("/loading.gif"));
+    private static final TileRetriever TILE_RETRIEVER = TileRetrieverProvider.getInstance().load();
+
+    public TileImageView(int zoom, long i, long j) {
+        setFitHeight(256);
+        setFitWidth(256);
+        setPreserveRatio(true);
+        setProgress(0);
+        CompletableFuture<Image> future = TILE_RETRIEVER.loadTile(zoom, i, j);
+        if (!future.isDone()) {
+            setImage(placeHolder);
```

Comment:
@eugener Any suggestion on how to expose this placeHolder image to the user?

Reply:
I wouldn't not use image property , but a function which returns image. This way it can be customized.
For example, we can pass a context, which includes the image size needed and that function would return appropriate image. We can then cache it. What do you think?

Reply:
Even if we add something like that to the API: 
```
public interface TileRetriever {

    CompletableFuture<Image> loadTile(int zoom, long i, long j);

   CompletableFuture<Image> loadTile(int zoom, long i, long j, Function<String, Image> placeholderFunction);

}
```

I don't see how we can expose that to the end user application? The placeholder image is used by `TileImageView`, which is the one that consumes internally the `TileRetriever`.

Reply:
What about the `ObjectProperty<Function>` on MapView? We might not even need to pass it on to the tile retriever - just wrap future's result into `Optional` and return cached placeholder if there is no image.

Reply:
If we define `ObjectProperty<Function>` on MapView, will need to pass it down to `BaseMap`, and then to `MapTile`, where we can create the `TileImageView` with the placeholder as an argument in its constructor. Does this make sense?

Reply:
One possible way to avoid this might be global "registration" of that function from within the `MapView` ( real property is static on some kind of internal class accessible to all other classes ). This has one limitation - no way to provide different placeholders to a two instances of `MapView`. But I don't think it is a problem?

Reply:
What about:

MapView:

```
private static Supplier<Image> placeholderImageSupplier;
public static Supplier<Image> getPlaceholderImageSupplier() { return placeholderImageSupplier; }
public static void setPlaceholderImageSupplier(Supplier<Image> supplier) { placeholderImageSupplier = supplier; }
```

and then in TileImageView:

```
public TileImageView(int zoom, long i, long j) {
        CompletableFuture<Image> future = TILE_RETRIEVER.loadTile(zoom, i, j);
        if (!future.isDone()) {
              // set placeholder image
              Optional.ofNullable(MapView.getPlaceholderImageSupplier()).ifPresent(s -> setImage(s.get()));
               ...
}
```

I don't think we need a property, the TileImageView constructor is called for every new tile.

From the end user app:

```
MapView mapView = new MapView();
MapView.setPlaceholderImageSupplier(() ->
                new Image(MapPresenter.class.getResourceAsStream("/loading.gif")));
 ...
```

Reply:
Looks good with one exception -  I would not tie `TileImageView` to the `MapView`.
I think `MapView.setPlaceholderImageSupplier` should set the `Supplier` to some internal class, possibly even to the `TileImageView` itself?

Reply:
Ok, pushed, please review.